### PR TITLE
feat: migrate wallet from json to sqlite

### PR DESCRIPTION
**Pull request by amirvalhalla**
## Description

Implementing sqlite for wallet.

### CODE REVIEW

**Comment by amirvalhalla** [comment]
please don't merge the branch. just review it.

File: wallet2/db/db.go [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+	}, nil
+}
+
+func (d *db) InsertIntoTransaction(t *Transaction) (*Transaction, error) {
+	insertQuery := fmt.Sprintf("INSERT INTO %s (tx_id, address, block_height, block_time,"+
+		" payload_type, data, description, amount, status, created_at) VALUES"+
+		" (?,?,?,?,?,?,?,?,?,?)", TransactionTable)
+
+	t.CreatedAt = time.Now().UTC()
+	r, err := d.ExecContext(context.Background(), insertQuery, t.TxID, t.Address, t.BlockHeight, t.BlockTime,
```
Comment (Ja7ad) [inline wallet2/db/db.go:170]: Please use top level context for prevent database damage on execute query.

```go
 InsertIntoTransaction(ctx context.Context, t *Transaction) (*Transaction, error)
```

File: wallet2/db/db.go [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+		PublicKey:  addr.PublicKey,
+		Label:      addr.Label,
+		Path:       addr.Path,
+		IsImported: addr.IsImported,
+		CreatedAt:  addr.CreatedAt,
+	}, nil
+}
+
+func (d *db) InsertIntoTransaction(t *Transaction) (*Transaction, error) {
+	insertQuery := fmt.Sprintf("INSERT INTO %s (tx_id, address, block_height, block_time,"+
```
Comment (Ja7ad) [inline wallet2/db/db.go:165]: Please don't use fmt.Sprintf for avoiding sql injection: https://go.dev/doc/database/sql-injection

File: wallet2/db/db.go (hunk 1 above)
Comment (Ja7ad) [inline wallet2/db/db.go:170]: Use prepared query before execute

**Review by Ja7ad — CHANGES_REQUESTED** [review]

File: wallet2/db/query.go [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package db
+
+import "fmt"
+
+type QueryOption string
+
+func WithTransactionStatus(status Status) QueryOption {
+	return QueryOption(fmt.Sprintf("WHERE status = %d", status))
```
Comment (Ja7ad) [inline wallet2/db/query.go:8]: don't use fmt.Sprintf

**Review by Ja7ad — CHANGES_REQUESTED** [review]

File: wallet2/db/query.go (hunk 3 above)
Comment (amirvalhalla) [reply wallet2/db/query.go:8]: what other option you suggest?

Comment (Ja7ad) [reply wallet2/db/query.go:8]: please check: https://go.dev/doc/database/sql-injection

Comment (Ja7ad) [reply wallet2/db/query.go:8]:
```go
func Query(query string, args ...any)
```

Comment (amirvalhalla) [reply wallet2/db/query.go:8]: this status comes from `HardCodec Code`, so doesn't make any security issue.

Comment (Ja7ad) [reply wallet2/db/query.go:8]: rows, err := db.Query("SELECT * FROM user WHERE id = ?", id)

File: wallet2/db/db.go (hunk 1 above)
Comment (Ja7ad) [reply wallet2/db/db.go:170]: https://pkg.go.dev/database/sql#DB.PrepareContext

File: wallet2/db/db.go (hunk 2 above)
Comment (amirvalhalla) [reply wallet2/db/db.go:165]: sql injection doesn't happen for this scenario because we're passing the table name from `hard coded section` and for other parameters we handled the sql injection.

File: wallet2/db/db.go [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+}
+
+type Pair struct {
+	Key       string
+	Value     string
+	CreatedAt time.Time
+}
+
+func NewDB(ctx context.Context, path string) (DB, error) {
+	dbInstance, err := sql.Open("sqlite", path)
```
Comment (Ja7ad) [inline wallet2/db/db.go:82]: Please create here 'context.WithCancel(ctx)' and keep cancel function in DB struct.
Then create Close method and call cancel func

**Review by Ja7ad — CHANGES_REQUESTED** [review]

File: wallet2/db/db.go (hunk 4 above)
Comment (amirvalhalla) [reply wallet2/db/db.go:82]: as we've talked before we stop it from wallet so the cancellation token will raise whenever wallet stop.

File: wallet2/db/db.go [hunk 5]
```diff
@@ -0,0 +1,549 @@
...
+	context.Context
+}
+
+type Address struct {
+	ID         int       `json:"id"`          // id of wallet
+	Address    string    `json:"address"`     // Address in the wallet
+	PublicKey  string    `json:"public_key"`  // Public key associated with the address
+	Label      string    `json:"label"`       // Label for the address
+	Path       string    `json:"path"`        // Path for the address
+	IsImported bool      `json:"is_imported"` // imported for purpose
```
Comment (b00f) [inline wallet2/db/db.go:57]: @amirvalhalla why we need to save "IsImported"?
We can get it from Path. right?

File: wallet2/db/db.go [hunk 6]
```diff
@@ -0,0 +1,549 @@
...
+	TxID        string    `json:"tx_id"`
+	Address     string    `json:"address"`
+	BlockHeight uint32    `json:"block_height"`
+	BlockTime   uint32    `json:"block_time"`
+	PayloadType string    `json:"payload_type"`
+	Data        string    `json:"data"`
+	Description string    `json:"description"`
+	Amount      int64     `json:"amount"`
+	Status      int       `json:"status"`
+	CreatedAt   time.Time `json:"created_at"`
```
Comment (b00f) [inline wallet2/db/db.go:72]: I think Sqlite automatically adds this column.

File: wallet2/db/db.go (hunk 5 above)
Comment (amirvalhalla) [reply wallet2/db/db.go:57]: Yes, we can. I just added this field for the future to help us in the GUI to display imported paths faster.

File: wallet2/db/db.go (hunk 6 above)
Comment (amirvalhalla) [reply wallet2/db/db.go:72]: I will delete json tags

Comment (b00f) [reply wallet2/db/db.go:72]: I mean "created_at" column.

File: wallet2/db/db.go (hunk 4 above)
Comment (Ja7ad) [reply wallet2/db/db.go:82]: We need to stop query when received interrupt signal.

For stop query you need send a cancellation signal via context cancel func.

File: wallet2/db/db.go (hunk 5 above)
Comment (b00f) [reply wallet2/db/db.go:57]: Is this the only reason?

File: wallet2/db/db.go [hunk 7]
```diff
@@ -0,0 +1,549 @@
+package db
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strings"
+	"time"
+
+	//nolint:all
```
Comment (b00f) [inline wallet2/db/db.go:10]: Remove this line and let fix the linting issue

File: wallet2/db/db.go [hunk 8]
```diff
@@ -0,0 +1,549 @@
...
+	_ "github.com/glebarez/go-sqlite"
+)
+
+const (
+	AddressTable     = "addresses"
+	TransactionTable = "transactions"
+	PairTable        = "pairs"
+)
+
+type DB interface {
```
Comment (b00f) [inline wallet2/db/db.go:20]: Do we need to define interface here?

File: wallet2/db/db.go [hunk 9]
```diff
@@ -0,0 +1,549 @@
...
+
+	GetAddressByID(id int) (*Address, error)
+	GetAddressByAddress(address string) (*Address, error)
+	GetAddressByPath(p string) (*Address, error)
+
+	GetTransactionByID(id int) (*Transaction, error)
+	GetTransactionByTxID(id string) (*Transaction, error)
+
+	GetPairByKey(key string) (*Pair, error)
+	GetTotalRecords(tableName string, query string, args ...any) (int64, error)
```
Comment (b00f) [inline wallet2/db/db.go:37]: The name is a bit confusing. 
If it returns the total record, why does it have a query parameter? 
So maybe we can rename it as "QueryTable."

File: wallet2/db/db.go [hunk 10]
```diff
@@ -0,0 +1,549 @@
...
+
+	GetAllAddresses() ([]Address, error)
+	GetAllAddressesWithTotalRecords(pageIndex, pageSize int) ([]Address, int64, error)
```
Comment (b00f) [inline wallet2/db/db.go:40]: A function can do many things is not recommended...

File: wallet2/db/db.go (hunk 9 above)
Comment (b00f) [reply wallet2/db/db.go:37]: Another confusion, why we have `GetTotalRecords` and `GetAllTransactions` 
They look doing the same thing

File: wallet2/db/db.go [hunk 11]
```diff
@@ -0,0 +1,549 @@
...
+	Description string    `json:"description"`
+	Amount      int64     `json:"amount"`
+	Status      int       `json:"status"`
+	CreatedAt   time.Time `json:"created_at"`
+}
+
+type Pair struct {
+	Key       string
+	Value     string
+	CreatedAt time.Time
```
Comment (b00f) [inline wallet2/db/db.go:78]: Same here, I believe SQLite handle it.

File: wallet2/db/db.go [hunk 12]
```diff
@@ -0,0 +1,549 @@
...
+	if err := d.createTransactionTable(); err != nil {
+		return err
+	}
+
+	return d.createPairTable()
+}
+
+func (d *db) createAddressTable() error {
+	addressQuery := fmt.Sprintf("CREATE TABLE %s (id INTEGER PRIMARY KEY AUTOINCREMENT,"+
+		" address VARCHAR, public_key VARCHAR, label VARCHAR, path VARCHAR, is_imported BOOLEAN, created_at TIMESTAMP)",
```
Comment (b00f) [inline wallet2/db/db.go:107]: Do we have fixed string size instead of VARCHAR in SQL?

File: wallet2/db/db.go [hunk 13]
```diff
@@ -0,0 +1,549 @@
...
+		AddressTable)
+
+	_, err := d.ExecContext(d.Context, addressQuery)
+	if err != nil && !strings.Contains(err.Error(), "already exists") {
```
Comment (b00f) [inline wallet2/db/db.go:111]: This code is not lovely! Any other way to check if a table exists?

File: wallet2/db/db.go [hunk 14]
```diff
@@ -0,0 +1,549 @@
...
+	GetAllAddresses() ([]Address, error)
+	GetAllAddressesWithTotalRecords(pageIndex, pageSize int) ([]Address, int64, error)
+
+	GetAllTransactions(query string, args ...any) ([]Transaction, error)
+	GetAllTransactionsWithTotalRecords(pageIndex, pageSize int, query string, args ...any) ([]Transaction, int64, error)
+}
+
+type db struct {
+	*sql.DB
+	context.Context
```
Comment (b00f) [inline wallet2/db/db.go:48]: To be same as other part of the code, let's give a name to it: `ctx`

File: wallet2/db/query.go [hunk 15]
```diff
@@ -0,0 +1,19 @@
+package db
+
+const (
+	EmptyQuery = ""
+)
+
+type QueryOption string
+
+func WithTransactionStatus() string {
```
Comment (b00f) [inline wallet2/db/query.go:9]: Maybe like
`func (qo *QueryOption) WithTransactionStatus() *QueryOption {...}`

**Review by b00f — CHANGES_REQUESTED** [review]

File: wallet2/db/db.go (hunk 6 above)
Comment (amirvalhalla) [reply wallet2/db/db.go:72]: I don't think so, let me check it.

File: wallet2/db/db.go (hunk 10 above)
Comment (amirvalhalla) [reply wallet2/db/db.go:40]: what function do you mean?

Comment (b00f) [reply wallet2/db/db.go:40]: `GetAllAddressesWithTotalRecords` It gets all address and calculates the total records...

File: wallet2/db/db.go (hunk 6 above)
Comment (amirvalhalla) [reply wallet2/db/db.go:72]: it doesn't automatically add `created_at` column

File: wallet2/db/db.go (hunk 8 above)
Comment (amirvalhalla) [reply wallet2/db/db.go:19]: it would be better to have, because we have some private methods which we don't want to expose them to higher layer of application plus for tests we don't have any problem because `newDB` method has input which we can pass `memory` then run all tests.

File: wallet2/db/db.go (hunk 9 above)
Comment (amirvalhalla) [reply wallet2/db/db.go:36]: `GetTotalRecords ` just returns the total count of records which they exist in db. we'll use it for pagination but `GetAllTransactions` returns all transaction records.

File: wallet2/db/db.go (hunk 10 above)
Comment (amirvalhalla) [reply wallet2/db/db.go:39]: I think for this case it's normal it doesn't make any issue. anyway we can remove total records from here and call directly `GetTotalRecords` method.

File: wallet2/db/db.go (hunk 11 above)
Comment (amirvalhalla) [reply wallet2/db/db.go:77]: the driver doesn't handle it

File: wallet2/db/db.go (hunk 12 above)
Comment (amirvalhalla) [reply wallet2/db/db.go:106]: we can use `VARCHAR(N)`, which `N` is size of string. should we refactor this section?

File: wallet2/db/db.go (hunk 13 above)
Comment (amirvalhalla) [reply wallet2/db/db.go:110]: I've searched for alternatives, but unfortunately, it seems there's no solution except for this one

File: wallet2/db/db.go (hunk 14 above)
Comment (amirvalhalla) [reply wallet2/db/db.go]: resolved.

File: wallet2/db/query.go (hunk 15 above)
Comment (amirvalhalla) [reply wallet2/db/query.go:9]: I think this way which we've used is better and I should delete `QueryOption` type. do you agree?

File: wallet2/db/db.go [hunk 16]
```diff
@@ -0,0 +1,549 @@
...
+}
+
+type Address struct {
```
Comment (b00f) [inline wallet2/db/db.go:50]: Why did you rename it to `Address`?
Don't you think `AddressInfo` make more sense?

File: wallet2/db/db.go (hunk 5 above)
Comment (b00f) [reply wallet2/db/db.go:56]: Anyway we can construct `AddressPath` once we reading from Database, then we can always check path for this purpose.

File: wallet2/db/db.go (hunk 12 above)
Comment (b00f) [reply wallet2/db/db.go:106]: `N` is size or max_size?

File: wallet2/db/db.go [hunk 17]
```diff
@@ -0,0 +1,549 @@
...
+	AddressTable     = "addresses"
+	TransactionTable = "transactions"
+	PairTable        = "pairs"
+)
+
+type DB interface {
+	CreateTables() error
+
+	InsertIntoAddress(addr *Address) (*Address, error)
+	InsertIntoTransaction(t *Transaction) (*Transaction, error)
```
Comment (b00f) [inline wallet2/db/db.go:23]: please avoid single letter for var names.

File: wallet2/db/db.go [hunk 18]
```diff
@@ -0,0 +1,549 @@
...
+	InsertIntoPair(key string, value string) (*Pair, error)
+
+	UpdateAddressLabel(addr *Address) (*Address, error)
+
+	GetAddressByID(id int) (*Address, error)
+	GetAddressByAddress(address string) (*Address, error)
+	GetAddressByPath(p string) (*Address, error)
```
Comment (b00f) [inline wallet2/db/db.go:30]: same here, no single letter

File: wallet2/db/db.go [hunk 19]
```diff
@@ -0,0 +1,549 @@
...
+
+	GetTransactionByID(id int) (*Transaction, error)
+	GetTransactionByTxID(id string) (*Transaction, error)
+
+	GetPairByKey(key string) (*Pair, error)
```
Comment (b00f) [inline wallet2/db/db.go:35]: Can we isolated Pair from outside of this package?
Then we can have `func GetValue(key string) (string, error)`

File: wallet2/db/db.go [hunk 20]
```diff
@@ -0,0 +1,549 @@
...
+type DB interface {
+	CreateTables() error
+
+	InsertIntoAddress(addr *Address) (*Address, error)
+	InsertIntoTransaction(t *Transaction) (*Transaction, error)
+	InsertIntoPair(key string, value string) (*Pair, error)
+
+	UpdateAddressLabel(addr *Address) (*Address, error)
+
+	GetAddressByID(id int) (*Address, error)
```
Comment (b00f) [inline wallet2/db/db.go:28]: I don't see any benefit by exporting ID or Index. Address string should be unique always.

**Review by b00f — CHANGES_REQUESTED** [review]

File: wallet2/db/db.go (hunk 12 above)
Comment (amirvalhalla) [reply wallet2/db/db.go:106]: N is max size. e.g. if we say `VARCHAR(50)` the maximum Len could be 50.

**Comment by Ja7ad** [comment]
@b00f Currently we don't need to migrate to sqlite and json worked.

Close it?

**Comment by b00f** [comment]
Yes, we can close it
